<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
</head>
<body>
  <script>
    const wait = (ms, ok) => {
      return new Promise((resolve, reject) => {
        setTimeout(() => {
          if (ok) {
            resolve(`過了 ${ms} 毫秒`)
          } else {
            reject(`錯誤，過了 ${ms} 毫秒`)
          }
        }, ms)
      })
    }

    // const aaa = wait(5000, true)
    // console.log(aaa)  // Promise<Pending>，因為 5 秒還沒到就 console.log

    // async function main () {}
    const main = async () => {
      try {
        // await 等待 Promise 結束，resolve 或 reject
        // 條件1: await 後面只能接 Promise
        // 條件2: await 只能放在 async function 裡面
        // async function 也是一種 Promise
        //
        // 變數 = await Promise
        // 變數會是 Promise resolve 的結果
        // 如果 Promise reject 會直接跳到 catch
        const result1 = await wait(1000, true)
        console.log(result1)
        const result2 = await wait(1000, false)
        console.log(result2)
        const result3 = await wait(1000, true)
        console.log(result3)
      } catch (error) {
        console.log('catch')
        console.log(error)
      }
    }
    main()
  </script>
</body>
</html>